File: 152-maximum_product_subarray.py
# time: O(n)
# space: O(1)
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        local_max = global_max = prev_min = prev_max = nums[0]

        for n in nums[1:]:
            local_min = min(prev_min * n, prev_max * n, n)
            local_max = max(prev_min * n, prev_max * n, n)

            global_max = max(local_max, global_max)
            
            prev_min = local_min
            prev_max = local_max

        return global_max

File: 152-maximum_product_subarray.cs
public class Solution {
    public int MaxProduct(int[] nums)
    {
        int lMax = nums[0];
        int lMin = lMax;
        int gMax = lMax;
        int pMin = lMax;
        int pMax = lMax;

        foreach(int n in nums.Skip(1))
        {
            lMax = new [] {n, n * pMin, n * pMax}.Max();
            lMin = new [] {n, n * pMin, n * pMax}.Min();

            pMax = lMax;
            pMin = lMin;

            gMax = Math.Max(lMax, gMax);
        }

        return gMax;
    }
}
